lib/vrl/compiler/src/enrichment/mod.rs

```diff
@@ -0,0 +1,54 @@
+use crate::Value;
+use dyn_clone::DynClone;
+use std::collections::BTreeMap;
+
+#[derive(Debug, PartialEq)]
+pub enum Condition<'a> {
+    Equals { field: &'a str, value: String },
+}
+
+pub trait TableSetup: DynClone {
+    fn table_ids(&self) -> Vec<String>;
+    fn add_index(&mut self, table: &str, fields: &[&str]) -> Result<(), String>;
+    fn as_readonly(&self) -> Box<dyn TableSearch + Send + Sync>;
+}
+
+dyn_clone::clone_trait_object!(TableSetup);
+
+pub trait TableSearch: DynClone + std::fmt::Debug {
+    fn find_table_row<'a>(
+        &'a self,
+        table: &str,
+        criteria: &'a [Condition<'a>],
+    ) -> Result<Option<BTreeMap<String, Value>>, String>;
+}
+
+dyn_clone::clone_trait_object!(TableSearch);
+
+/// Create a empty enrichment for situations when we don't have any tables loaded.
+#[derive(Clone, Debug)]
+pub struct EmptyEnrichmentTables;
```

Comment:
As mentioned offline: I think this should be done through an `impl<T: TableSetup> for Option<T>`, but we dan do that in a follow-up PR. Just noting this so we don't forget to track that work.